include/utils/actions/smooth.h

```diff
@@ -0,0 +1,151 @@
+
+        return std::make_tuple(AB_magnitude, BC_magnitude, CD_magnitude);
+    }
+
+    template<class Point>
+    requires utils::point2d<Point> || utils::junction<Point>
+    constexpr auto cosAngle(Point* A, Point* B, Point* C, const std::floating_point auto AB_magnitude, const std::floating_point auto BC_magnitude) const noexcept
+    {
+        if (AB_magnitude == 0.0 || BC_magnitude == 0.0)
+        {
+            return 0.0;
+        }
+        auto AB = std::make_tuple(std::get<"X">(*B) - std::get<"X">(*A), std::get<"Y">(*B) - std::get<"Y">(*A));
+        auto BC = std::make_tuple(std::get<"X">(*C) - std::get<"X">(*B), std::get<"Y">(*C) - std::get<"Y">(*B));
+
+        const auto dot = dotProduct(&AB, &BC);
+        return dot / (AB_magnitude * BC_magnitude);
+    }
+
+    template<class Point>
+    requires utils::point2d<Point> || utils::junction<Point>
+    constexpr void shiftPointTowards(Point* point, Point* target, const std::floating_point auto p0p1_distance, const std::integral auto smooth_distance) const noexcept
+    {
+        using coord_type = std::remove_cvref_t<decltype(std::get<"X">(*point))>;
+        const auto shift_distance = smooth_distance / p0p1_distance;
+        const auto shift_distance_x = static_cast<coord_type>((std::get<"X">(*target) - std::get<"X">(*point)) * shift_distance);
+        const auto shift_distance_y = static_cast<coord_type>((std::get<"Y">(*target) - std::get<"Y">(*point)) * shift_distance);
+        if constexpr (utils::point2d<Point>)
+        {
+            point->X += shift_distance_x;
+            point->Y += shift_distance_y;
+        }
+        else
+        {
+            point->p.X += shift_distance_x;
+            point->p.Y += shift_distance_y;
+        }
+    }
+
+    template<class Vector>
+        requires utils::point2d<Vector> || utils::junction<Vector> constexpr auto dotProduct(Vector* point_0, Vector* point_1) const noexcept
+    {
+        return std::get<"X">(*point_0) * std::get<"X">(*point_1) + std::get<"Y">(*point_0) * std::get<"Y">(*point_1);
+    }
+
+    template<class Point>
+    requires utils::point2d<Point> || utils::junction<Point>
+    constexpr auto isWithinAllowedDeviations(Point* A, Point* B, Point* C, Point* D, const std::floating_point auto fluid_angle, const std::integral auto max_resolution,
+                                          const std::floating_point auto AB_magnitude, const std::floating_point auto BC_magnitude, const std::floating_point auto CD_magnitude) const noexcept
+    {
+        if (BC_magnitude > max_resolution / 10) // TODO: make dedicated front-end setting for this
+        {
+            return true;
+        }
+        const double cos_A = std::acos(cosAngle(A, B, C, AB_magnitude, BC_magnitude));
+        const double cos_B = std::acos(cosAngle(A, B, D, AB_magnitude, CD_magnitude));
+        const auto abs_angle = std::abs(cos_A - cos_B);
+        return abs_angle < fluid_angle;
```

Comment:
Why calculate the angle `angle(ab, bc)` and `angle(ab, bd)` independently instead of directly check the angle `angle(ab, cd)`?

Pseudo code would be something like
```
const auto ab = b - a;
const auto cd = d - c;
const cos_angle = dot(ab, bc) / (length(ab) * length(bc));
const is_sharp_angle = cos_angle < cos(fluid_angle);
```
Not that this reduces the number of `acos`/`cos` calculations dramastically as `cos(fluid_angle)` is constant; you only need to calculate it once.

Reply:
The idea behind this smoothing is that the angle of BC should be within a fluid motion path, from where it comes from to where it will be going.

Reply:
Lets discuss behind a whiteboard

Reply:
Put this on the C&C board to be discussed, for now the current check is acceptable